File: packages/sdk/rust/packages/admin/src/analytics.rs
use std::collections::HashMap;
use std::sync::Arc;
use std::time::{SystemTime, UNIX_EPOCH};

use edgebase_core::error::Error;
use edgebase_core::generated::api_core::GeneratedDbApi;
use edgebase_core::http_client::HttpClient;
use serde_json::{json, Value};

use crate::generated::admin_api_core::GeneratedAdminApi;

pub struct AnalyticsClient {
    http: Arc<HttpClient>,
}

impl AnalyticsClient {
    pub(crate) fn new(http: Arc<HttpClient>) -> Self {
        Self { http }
    }

    pub async fn overview(&self, options: Option<HashMap<String, String>>) -> Result<Value, Error> {
        self.admin_core()
            .query_analytics(&self.build_query("overview", options))
            .await
    }

    pub async fn time_series(&self, options: Option<HashMap<String, String>>) -> Result<Vec<Value>, Error> {
        let result = self
            .admin_core()
            .query_analytics(&self.build_query("timeSeries", options))
            .await?;
        Ok(extract_list(&result, "timeSeries"))
    }

    pub async fn breakdown(&self, options: Option<HashMap<String, String>>) -> Result<Vec<Value>, Error> {
        let result = self
            .admin_core()
            .query_analytics(&self.build_query("breakdown", options))
            .await?;
        Ok(extract_list(&result, "breakdown"))
    }

    pub async fn top_endpoints(&self, options: Option<HashMap<String, String>>) -> Result<Vec<Value>, Error> {
        let result = self
            .admin_core()
            .query_analytics(&self.build_query("topEndpoints", options))
            .await?;
        Ok(extract_list(&result, "topItems"))
    }

    pub async fn track(
        &self,
        name: &str,
        properties: Option<Value>,
        user_id: Option<&str>,
    ) -> Result<(), Error> {
        let mut event = serde_json::Map::new();
        event.insert("name".to_string(), Value::String(name.to_string()));
        event.insert("timestamp".to_string(), json!(current_time_millis()));
        if let Some(props) = properties {
            event.insert("properties".to_string(), props);
        }
        if let Some(user) = user_id {
            event.insert("userId".to_string(), Value::String(user.to_string()));
        }
        self.track_batch(vec![Value::Object(event)]).await
    }

    pub async fn track_batch(&self, events: Vec<Value>) -> Result<(), Error> {
        if events.is_empty() {
            return Ok(());
        }

        let normalized = events
            .into_iter()
            .map(|event| {
                if let Value::Object(mut map) = event {
                    map.entry("timestamp".to_string())
                        .or_insert_with(|| json!(current_time_millis()));
                    Value::Object(map)
                } else {
                    event
                }
            })
            .collect::<Vec<_>>();

        self.core()
            .track_events(&json!({ "events": normalized }))
            .await?;
        Ok(())
    }

    pub async fn query_events(&self, options: Option<HashMap<String, String>>) -> Result<Value, Error> {
        let params = options.unwrap_or_default();
        self.admin_core().query_custom_events(&params).await
    }

    fn core(&self) -> GeneratedDbApi<'_> {
        GeneratedDbApi::new(&self.http)
    }

    fn admin_core(&self) -> GeneratedAdminApi<'_> {
        GeneratedAdminApi::new(&self.http)
    }

    fn build_query(&self, metric: &str, options: Option<HashMap<String, String>>) -> HashMap<String, String> {
        let mut query = HashMap::from([("metric".to_string(), metric.to_string())]);
        if let Some(options) = options {
            query.extend(options);
        }
        query
    }
}

fn extract_list(result: &Value, field: &str) -> Vec<Value> {
    result
        .get(field)
        .and_then(|value| value.as_array())
        .cloned()
        .unwrap_or_default()
}

fn current_time_millis() -> u128 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|duration| duration.as_millis())
        .unwrap_or(0)
}

File: packages/sdk/js/packages/admin/src/analytics.ts
/**
 * AnalyticsClient — Analytics query + custom event tracking for Admin/Server SDK
 *
 * Feature 1: Request log metrics query (same data as admin dashboard)
 * Feature 2: Custom event tracking (server-side) + query
 *
 * Usage:
 *   const admin = createAdminClient(url, { serviceKey });
 *
 *   // Request log metrics
 *   const overview = await admin.analytics.overview({ range: '7d' });
 *   const ts = await admin.analytics.timeSeries({ range: '24h', category: 'db' });
 *
 *   // Custom events
 *   await admin.analytics.track('user_upgraded', { plan: 'pro', amount: 29.99 }, 'user-123');
 *   const events = await admin.analytics.queryEvents({ event: 'user_upgraded', metric: 'count' });
 */
import type { HttpClient } from '@edgebase-fun/core';
import { HttpClientAdapter, DefaultDbApi } from '@edgebase-fun/core';
import { DefaultAdminApi } from './generated/admin-api-core.js';

// ─── Feature 1: Request Log Metric Types ───

/** Options for querying request log metrics. */
export interface AnalyticsQueryOptions {
  /** Time range for the query. */
  range?: '1h' | '6h' | '24h' | '7d' | '30d' | '90d';
  /** Filter by route category (e.g. 'db', 'auth', 'storage'). */
  category?: string;
  /** Grouping interval for time series data. */
  groupBy?: 'minute' | 'hour' | 'day';
}

/** A single point in a time series. */
export interface TimeSeriesPoint {
  timestamp: number;
  requests: number;
  errors: number;
  avgLatency: number;
  uniqueUsers: number;
}

/** Aggregate summary of request metrics. */
export interface AnalyticsSummary {
  totalRequests: number;
  totalErrors: number;
  avgLatency: number;
  uniqueUsers: number;
}

/** Breakdown item (e.g. by category or status code). */
export interface BreakdownItem {
  label: string;
  count: number;
  percentage: number;
  avgLatency?: number;
  errorRate?: number;
}

/** Top endpoint/item entry. */
export interface TopItem {
  label: string;
  count: number;
  avgLatency: number;
  errorRate: number;
}

/** Full overview response from the analytics query endpoint. */
export interface AnalyticsOverview {
  timeSeries: TimeSeriesPoint[];
  summary: AnalyticsSummary;
  breakdown: BreakdownItem[];
  topItems: TopItem[];
}

// ─── Feature 2: Custom Event Types ───

/** Data for tracking a single custom event. */
export interface TrackEventData {
  /** Event name (required). */
  name: string;
  /** Arbitrary properties (max 50 keys, max 4KB JSON). */
  properties?: Record<string, string | number | boolean>;
  /** Unix timestamp in ms (default: now). */
  timestamp?: number;
  /** User ID override (Service Key only — ignored with JWT). */
  userId?: string;
}

/** Options for querying custom events. */
export interface EventQueryOptions {
  /** Time range. */
  range?: '1h' | '6h' | '24h' | '7d' | '30d' | '90d';
  /** Filter by event name. */
  event?: string;
  /** Filter by user ID. */
  userId?: string;
  /** Query metric type. */
  metric?: 'list' | 'count' | 'timeSeries' | 'topEvents';
  /** Grouping interval for time series. */
  groupBy?: 'minute' | 'hour' | 'day';
  /** Max results (for list metric). */
  limit?: number;
  /** Cursor for pagination (for list metric). */
  cursor?: string;
}

/** A single custom event record. */
export interface EventItem {
  id: number;
  timestamp: number;
  userId: string | null;
  eventName: string;
  properties: Record<string, unknown> | null;
}

/** Result for metric='list'. */
export interface EventListResult {
  events: EventItem[];
  cursor?: string;
  hasMore: boolean;
}

/** Result for metric='count'. */
export interface EventCountResult {
  totalEvents: number;
  uniqueUsers: number;
}

/** Result for metric='timeSeries'. */
export interface EventTimeSeriesResult {
  timeSeries: Array<{ timestamp: number; count: number }>;
}

/** Result for metric='topEvents'. */
export interface EventTopResult {
  topEvents: Array<{ eventName: string; count: number; uniqueUsers: number }>;
}

// ─── AnalyticsClient ───

export class AnalyticsClient {
  private core: DefaultDbApi;
  private adminCore: DefaultAdminApi;

  constructor(httpClient: HttpClient) {
    const adapter = new HttpClientAdapter(httpClient);
    this.core = new DefaultDbApi(adapter);
    this.adminCore = new DefaultAdminApi(adapter);
  }

  // ── Feature 1: Request Log Metrics ──

  /**
   * Get a full analytics overview (time series + summary + breakdown + top endpoints).
   *
   * @example
   * const overview = await admin.analytics.overview({ range: '7d' });
   * console.log(overview.summary.totalRequests);
   */
  async overview(options?: AnalyticsQueryOptions): Promise<AnalyticsOverview> {
    const query = this.buildQuery({ ...options, metric: 'overview' });
    return this.adminCore.queryAnalytics(query) as Promise<AnalyticsOverview>;
  }

  /**
   * Get time series data only.
   *
   * @example
   * const ts = await admin.analytics.timeSeries({ range: '24h', category: 'db' });
   */
  async timeSeries(options?: AnalyticsQueryOptions): Promise<TimeSeriesPoint[]> {
    const query = this.buildQuery({ ...options, metric: 'timeSeries' });
    const res = await this.adminCore.queryAnalytics(query) as { timeSeries: TimeSeriesPoint[] };
    return res.timeSeries;
  }

  /**
   * Get breakdown data (by category, status code, etc.).
   *
   * @example
   * const breakdown = await admin.analytics.breakdown({ range: '30d' });
   */
  async breakdown(options?: AnalyticsQueryOptions): Promise<BreakdownItem[]> {
    const query = this.buildQuery({ ...options, metric: 'breakdown' });
    const res = await this.adminCore.queryAnalytics(query) as { breakdown: BreakdownItem[] };
    return res.breakdown;
  }

  /**
   * Get top endpoints by request count.
   *
   * @example
   * const top = await admin.analytics.topEndpoints({ range: '7d', category: 'auth' });
   */
  async topEndpoints(options?: AnalyticsQueryOptions): Promise<TopItem[]> {
    const query = this.buildQuery({ ...options, metric: 'topEndpoints' });
    const res = await this.adminCore.queryAnalytics(query) as { topItems: TopItem[] };
    return res.topItems;
  }

  // ── Feature 2: Custom Events ──

  /**
   * Track a single custom event.
   *
   * @param name       Event name (e.g. 'user_upgraded')
   * @param properties Arbitrary key-value data (max 50 keys, max 4KB)
   * @param userId     User ID to associate (Service Key callers only)
   *
   * @example
   * await admin.analytics.track('user_upgraded', { plan: 'pro', amount: 29.99 }, 'user-123');
   */
  async track(
    name: string,
    properties?: Record<string, string | number | boolean>,
    userId?: string,
  ): Promise<void> {
    const event: Record<string, unknown> = { name };
    if (properties) event.properties = properties;
    if (userId) event.userId = userId;
    await this.core.trackEvents({ events: [event] });
  }

  /**
   * Track multiple custom events in a single request (max 100).
   *
   * @example
   * await admin.analytics.trackBatch([
   *   { name: 'page_view', properties: { path: '/pricing' } },
   *   { name: 'page_view', properties: { path: '/docs' } },
   * ]);
   */
  async trackBatch(events: TrackEventData[]): Promise<void> {
    await this.core.trackEvents({ events });
  }

  /**
   * Query custom events. Returns different shapes based on `metric`.
   *
   * @example
   * // List events
   * const list = await admin.analytics.queryEvents({ event: 'purchase', metric: 'list', limit: 20 });
   *
   * // Count
   * const count = await admin.analytics.queryEvents({ event: 'purchase', metric: 'count' });
   *
   * // Time series
   * const ts = await admin.analytics.queryEvents({ metric: 'timeSeries', groupBy: 'day' });
   *
   * // Top events
   * const top = await admin.analytics.queryEvents({ metric: 'topEvents' });
   */
  async queryEvents<T = EventListResult | EventCountResult | EventTimeSeriesResult | EventTopResult>(
    options?: EventQueryOptions,
  ): Promise<T> {
    const query: Record<string, string> = {};
    if (options?.range) query.range = options.range;
    if (options?.event) query.event = options.event;
    if (options?.userId) query.userId = options.userId;
    if (options?.metric) query.metric = options.metric;
    if (options?.groupBy) query.groupBy = options.groupBy;
    if (options?.limit) query.limit = String(options.limit);
    if (options?.cursor) query.cursor = options.cursor;
    return this.adminCore.queryCustomEvents(query) as Promise<T>;
  }

  // ── Internal ──

  private buildQuery(opts: Partial<AnalyticsQueryOptions> & { metric: string }): Record<string, string> {
    const q: Record<string, string> = { metric: opts.metric };
    if (opts.range) q.range = opts.range;
    if (opts.category) q.category = opts.category;
    if (opts.groupBy) q.groupBy = opts.groupBy;
    return q;
  }
}
